libkbfs/block_tree.go

```diff
@@ -269,179 +275,235 @@ func (bt *blockTree) getNextDirtyBlockAtOffset(ctx context.Context,
+			pblock:     pblock,
+			childIndex: childIndex,
+		}
 
-			// Append self to the front of every path.
-			var r resp
-			for _, p := range pfr {
-				newPath := append([]parentBlockAndChildIndex{{
-					pblock:     pblock,
-					childIndex: childIndex,
-				}}, p...)
-				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+		// We only need to fetch direct blocks if we've been asked
+		// to do so.  If the direct type of the pointer is
+		// unknown, we can assume all the children are direct
+		// blocks, since there weren't multiple levels of
+		// indirection before the introduction of the flag.
+		if job.getDirect || childPtr.DirectType == IndirectBlock {
+			// Create a task for the child, and either enqueue it or recurse to it.
+
+			subTask := job.subTask(childPtr, childPath)
+
+			// Enqueue the subTask with the WorkerPool.
+			wg.Add(1)
+			wp.Submit(func() {
+				bt.processGetBlocksTask(ctx, wg, wp, subTask, results)
+			})
+		} else {
+			results <- getBlocksForOffsetRangeResult{
+				pathFromRoot: childPath,
 			}
-			r.blocks = blocks
-			r.nextBlockOffset = nextBlockOffset
-			respCh <- r
-			return nil
-		})
+		}
 	}
+}
 
-	err = eg.Wait()
-	// If we are ok with just getting the prefix, don't treat a
-	// deadline exceeded error as fatal.
-	if prefixOk && err == context.DeadlineExceeded {
-		err = nil
-	}
-	if err != nil {
-		return nil, nil, nil, err
+// getBlocksForOffsetRange fetches all the blocks making up paths down
+// the block tree to leaf ("direct") blocks that encompass the given
+// offset range (half-inclusive) in the data.  If `endOff` is nil, it
+// returns blocks until reaching the end of the data.  If `prefixOk`
+// is true, the function will ignore context deadline errors and
+// return whatever prefix of the data it could fetch within the
+// deadine.  Return params:
+//
+//   * pathsFromRoot is a slice, ordered by offset, of paths from
+//     the root to each block that makes up the range.  If the path is
+//     empty, it indicates that pblock is a direct block and has no
+//     children.
+//   * blocks: a map from block pointer to a data-containing leaf node
+//     in the given range of offsets, if `getDirect` is true.
+//   * nextBlockOff is the offset of the block that follows the last
+//     block given in `pathsFromRoot`.  If `pathsFromRoot` contains
+//     the last block among the children, nextBlockOff is nil.
+func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
+	ptr BlockPointer, pblock BlockWithPtrs, startOff, endOff Offset,
+	prefixOk bool, getDirect bool) (pathsFromRoot [][]parentBlockAndChildIndex,
+	blocks map[BlockPointer]Block, nextBlockOffset Offset,
+	err error) {
+	// Make a WaitGroup to keep track of whether there's still work to be done.
+	var wg sync.WaitGroup
+
+	// Make a workerpool to limit the number of concurrent goroutines.
+	wp := workerpool.New(maxBlockFetchWorkers)
+	defer wp.Stop() // TODO: this may be unnecessary
```

Comment:
Unless you think a "TODO" requires further ticketed work, please address them and remove the comments before merging.  If they do require more work, make a ticket for them in JIRA and add the ticket number here, like `// TODO(KBFS-3503): ...`.

In this case, this is definitely not needed because you make the same call again in a later goroutine.  But in general, the workerpool documentation implies to me that the `Stop` call is needed to clean up the goroutines launched by it (though I haven't checked the code).